src/main/java/techreborn/compat/libcd/TRRecipeParser.java

```diff
@@ -0,0 +1,102 @@
+                amount = Optional.of(Integer.parseInt(count));
+                in = in.substring(0, atIndex);
+            }
+            if (in.indexOf('~') == 0) {
+                //fluids
+                Identifier id = new Identifier(in.substring(1));
+                Fluid fluid = Registry.FLUID.get(id);
+                if (fluid == Fluids.EMPTY) throw new TweakerSyntaxException("Failed to get fluid for input: " + in);
+                return new FluidIngredient(fluid, Optional.empty(), amount);
+            } else if (in.indexOf('#') == 0) {
+                Identifier id = new Identifier(in.substring(1));
+                Tag<Item> itemTag = ItemTags.getContainer().get(id);
+                if (itemTag == null) throw new TweakerSyntaxException("Failed to get item tag for input: " + in);
+                return new TagIngredient(id, ItemTags.getContainer().get(id), amount);
+            } else {
+                ItemStack stack;
+                Optional<CompoundTag> tag = Optional.empty();
+                boolean requireEmpty = false;
+                if (in.contains("->")) {
+                    ItemStack readStack = TweakerUtils.INSTANCE.getSpecialStack(in);
+                    if (readStack.isEmpty())
+                        throw new TweakerSyntaxException("Failed to get special stack for input: " + in);
+                    if (readStack.hasTag()) {
+                        tag = Optional.of(readStack.getTag());
+                    } else {
+                        requireEmpty = true;
+                    }
+                    stack = readStack;
+                } else {
+                    int nbtIndex = in.indexOf('{');
+                    if (nbtIndex != -1) {
+                        try {
+                            String nbt = in.substring(nbtIndex);
+                            in = in.substring(0, nbtIndex);
+                            StringNbtReader reader = new StringNbtReader(new StringReader(nbt));
+                            CompoundTag parsedTag = reader.parseCompoundTag();
+                            if (parsedTag.isEmpty()) requireEmpty = true;
+                            else tag = Optional.of(reader.parseCompoundTag());
+                        } catch (CommandSyntaxException e) {
+                            throw new TweakerSyntaxException(e.getMessage());
+                        }
+                    }
+                    Identifier id = new Identifier(in);
+                    Item item = Registry.ITEM.get(id);
+                    if (item == Items.AIR) throw new TweakerSyntaxException("Failed to get item for input: " + in);
+                    stack = new ItemStack(item);
+                }
+                return new StackIngredient(Collections.singletonList(stack), amount, tag, requireEmpty);
+            }
+        }
+        else throw new TweakerSyntaxException("Illegal object passed to TechReborn of type " + input.getClass().getName());
+    }
```

Comment:
Quite a bit of code here that seems like it should be in libcd. Is there a specific reason why this needs to be in TR? (Thinking of the maintenance burden)

Reply:
This *is* effectively in LibCD already, but TR needs its own copy because it doesn't use vanilla ingredients.

Reply:
Humm yeah, good point. What about using the wrapped indgredient? If not this isnt too much.

Reply:
The primary issue is that vanilla ingredients only accept one item at a time; you can't set a count like you can in TR. That's a big problem with vanilla in general, but I'm not sure how feasible it would be to have LibCD mix into all the vanilla stuff to fix it.

Reply:
Ah ok thats a good point. Dont worry about it its not like its 1000s of lines.